ocaml/xapi-guard/lib/disk_cache.ml

```diff
@@ -0,0 +1,596 @@
+   MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
+   GNU Lesser General Public License for more details.
+*)
+
+module D = Debug.Make (struct let name = __MODULE__ end)
+
+let ( // ) = Filename.concat
+
+let runtime_data = "/var/lib" // "xapi-guard"
+
+let ( let* ) = Lwt.bind
+
+let ( let@ ) f x = f x
+
+let with_lock = Lwt_mutex.with_lock
+
+type t = Uuidm.t * Mtime.t * Types.Tpm.key
+
+let cache_of service = runtime_data // Types.Service.to_string service
+
+let fistpoint () =
+  let name = "/tmp/fist_disable_xapi_guard_cache" in
+  Lwt.catch
+    (fun () ->
+      let* () = Lwt_unix.access name [Unix.F_OK] in
+      Lwt.return true
+    )
+    (fun _ -> Lwt.return false)
+
+let files_in dir ~otherwise =
+  Lwt.catch
+    (fun () ->
+      let* listing = Lwt_unix.files_of_directory dir |> Lwt_stream.to_list in
+      List.filter_map
+        (function "." | ".." -> None | name -> Some (dir // name))
+        listing
+      |> Lwt.return
+    )
+    otherwise
+
+let unlink_safe file =
+  let __FUN = __FUNCTION__ in
+  Lwt.catch
+    (fun () -> Lwt_unix.unlink file)
+    (function
+      | Unix.(Unix_error (ENOENT, _, _)) ->
+          Lwt.pause ()
+      | e ->
+          D.info "%s: error %s when deleting %s, ignoring" __FUN
+            (Printexc.to_string e) file ;
+          Lwt.pause ()
+      )
+
+type valid_file = t * string
+
+type file =
+  | Latest of valid_file
+  | Outdated of valid_file
+  | Temporary of string
+  | Invalid of string
+
+let path_of_key root (uuid, timestamp, key) =
+  root
+  // Uuidm.to_string uuid
+  // Types.Tpm.(serialize_key key |> string_of_int)
+  // Mtime.(to_uint64_ns timestamp |> Int64.to_string)
+
+let key_of_path path =
+  let ( let* ) = Option.bind in
+  let key_dir = Filename.(dirname path) in
+  let* uuid = Filename.(basename (dirname key_dir)) |> Uuidm.of_string in
+  let* key =
+    Filename.basename key_dir
+    |> int_of_string_opt
+    |> Option.map Types.Tpm.deserialize_key
+  in
+  let* timestamp =
+    Filename.basename path
+    |> Int64.of_string_opt
+    |> Option.map Mtime.of_uint64_ns
+  in
+  Some ((uuid, timestamp, key), path)
+
+let path_is_temp path =
+  let pathlen = String.length path in
+  String.ends_with ~suffix:".pre" path
+  && key_of_path (String.sub path 0 (pathlen - 4)) |> Option.is_some
+
+let temp_of_path path = path ^ ".pre"
+
+let sort_updates contents =
+  let rec loop (acc, found) = function
+    | [] ->
+        List.rev acc
+    | latest :: others -> (
+      match key_of_path latest with
+      | None ->
+          let file =
+            if path_is_temp latest then
+              Temporary latest
+            else
+              Invalid latest
+          in
+          loop (file :: acc, found) others
+      | Some valid_file ->
+          let file = if found then Outdated valid_file else Latest valid_file in
+          loop (file :: acc, true) others
+    )
+  in
+  let ordered = List.fast_sort (fun x y -> String.compare y x) contents in
```

Comment:
Are we sure this is the correct comparison function? Looking at `let filename` below it uses Int64.to_string, which means that smaller values are shorter strings and larger values are longer strings, instead of using a fixed width padded with zeros.

For example try sorting this list and the results are not what I'd call sorted:
```
List.fast_sort (fun x y -> String.compare y x) ["1";"9";"10";"90";"100"];;
- : string list = ["90"; "9"; "100"; "10"; "1"]
```

In fact this means that we'll lose the actually latest value, and potentially overwrite it with an older value, depending on when the Mtime changes length when converted to a string.

To fix this either convert the string back to an integer and compare it (but that is error prone, what if in the future we add some non-numeric characters there), *or* add some padding when constructing the string (e.g. `Printf.sprintf "%08xL"` instead of `Int64.to_string`).

In the future we probably want to remove the need for sorting and keep just one file, which would avoid bugs like these (but that is a larger change to do now).

Reply:
Please check the latest commit